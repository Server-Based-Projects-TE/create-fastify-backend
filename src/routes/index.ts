export * from './basicRoutes';
export * from './usersRoutes';
