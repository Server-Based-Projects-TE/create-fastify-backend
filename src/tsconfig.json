{
  "extends": "./../node_modules/@tsconfig/node14/tsconfig.json",
  "compilerOptions": {
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "declaration": true,
    "rootDir": ".",
    "outDir": "../build/",
    "baseUrl": "./",
    "paths": {
      "src/*": ["./*"]
    }
  },
  "references": [{ "path": "../" }]
}
