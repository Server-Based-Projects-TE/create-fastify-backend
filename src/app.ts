import fastifySession, { MemoryStore } from '@mgcrea/fastify-session';
import { RedisStore } from '@mgcrea/fastify-session-redis-store';
import createFastify, { FastifyInstance, FastifyServerOptions } from 'fastify';
import fastifyCookie from 'fastify-cookie';
import fastifyPassport from 'fastify-passport';
import Redis from 'ioredis';
import { IS_PROD, IS_TEST, REDIS_URI, SESSION_KEY, SESSION_TTL } from './config/env';
import { DEFAULT_LOGGER } from './config/logger';
import { LocalStrategy } from './config/passport/LocalStrategy';
import { deserializeFromSession, serializeIntoSession } from './config/passport/session';
import { fastifyAuth, FastifyAuthPluginOptions } from './plugins/auth';
import { baseRoutes, sessionsRoutes, usersRoutes } from './routes';

export type BuildFastifyOptions = FastifyServerOptions & FastifyAuthPluginOptions;

export const buildFastify = (options: BuildFastifyOptions = {}): FastifyInstance => {
  const { logger = DEFAULT_LOGGER, ...otherOptions } = options;
  const fastify = createFastify({ logger, ...otherOptions });

  // Support sessions using cookies
  fastify.register(fastifyCookie);
  fastify.register(fastifySession, {
    store: IS_TEST ? new MemoryStore() : new RedisStore({ client: new Redis(REDIS_URI), ttl: SESSION_TTL }),
    key: SESSION_KEY,
    cookie: { secure: IS_PROD ? true : false, maxAge: SESSION_TTL },
  });

  // Initialize passport
  fastify.register(fastifyPassport.initialize());
  // - with support for sessions
  fastify.register(fastifyPassport.secureSession());
  fastifyPassport.registerUserSerializer(serializeIntoSession);
  fastifyPassport.registerUserDeserializer(deserializeFromSession);
  // - with specific auth strategies
  fastifyPassport.use(new LocalStrategy());
  // - and auth helper plugin
  fastify.register(fastifyAuth, {
    bypassAuth: options.bypassAuth,
  });

  // Register routes
  fastify.register(baseRoutes);
  fastify.register(sessionsRoutes);
  fastify.register(usersRoutes);

  return fastify;
};
