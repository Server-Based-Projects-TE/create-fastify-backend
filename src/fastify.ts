import createFastify from 'fastify';
import { logger } from './config/logger';
import { basicRoutes, usersRoutes } from './routes';

export const fastify = createFastify({
  logger,
});

export const log = fastify.log;

fastify.register(basicRoutes);
fastify.register(usersRoutes);
