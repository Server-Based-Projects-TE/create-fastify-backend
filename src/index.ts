import { buildFastify } from './app';
import { NODE_HOST, NODE_PORT } from './config/env';
import './config/setup';

const server = buildFastify();

server.listen(NODE_PORT, NODE_HOST, (err) => {
  if (err) {
    server.log.error(`${err.message}\n${err.stack}`);
    process.exit(1);
  }
});
