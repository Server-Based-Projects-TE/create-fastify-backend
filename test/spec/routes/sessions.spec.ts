import { buildFastify } from 'src/app';

describe('routes/sessions', () => {
  const fastify = buildFastify();
  const context = new Map();
  afterAll(async () => {
    fastify.close();
  });
  it('should receive a 401 on GET /me without auth', async () => {
    const response = await fastify.inject({
      method: 'GET',
      url: '/me',
    });
    expect(response.statusCode).toEqual(401);
    expect(response.json()).toMatchSnapshot();
  });
  it('should receive a 400 on POST /login with missing credentials', async () => {
    const response = await fastify.inject({
      method: 'POST',
      payload: { username: 'foo' },
      url: '/login',
    });
    expect(response.statusCode).toEqual(400);
    expect(response.json()).toMatchSnapshot();
  });
  it('should receive a 403 on POST /login with invalid credentials', async () => {
    const response = await fastify.inject({
      method: 'POST',
      payload: { username: 'foo', password: 'qux' },
      url: '/login',
    });
    expect(response.statusCode).toEqual(403);
    expect(response.json()).toMatchSnapshot();
  });
  it('should receive a 200 on POST /login with valid credentials', async () => {
    const response = await fastify.inject({
      method: 'POST',
      payload: { username: 'foo', password: 'bar' },
      url: '/login',
    });
    expect(response.statusCode).toEqual(200);
    expect(response.json()).toMatchSnapshot();
    expect(Object.keys(response.headers)).toContain('set-cookie');
    context.set('cookie', response.headers['set-cookie']);
  });
  it('should receive a 200 on GET /me with a valid session', async () => {
    const response = await fastify.inject({
      method: 'GET',
      url: '/me',
      headers: {
        cookie: context.get('cookie').split(';')[0],
      },
    });
    expect(response.statusCode).toEqual(200);
    expect(response.json()).toMatchSnapshot();
    expect(Object.keys(response.headers)).not.toContain('set-cookie');
  });
});
