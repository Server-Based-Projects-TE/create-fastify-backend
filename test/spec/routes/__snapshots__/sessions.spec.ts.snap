// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/sessions should receive a 200 on GET /me with a valid session 1`] = `
Object {
  "id": "1",
  "username": "foo",
}
`;

exports[`routes/sessions should receive a 200 on POST /login with valid credentials 1`] = `
Object {
  "ok": 1,
  "user": Object {
    "id": "1",
    "username": "foo",
  },
}
`;

exports[`routes/sessions should receive a 400 on POST /login with missing credentials 1`] = `
Object {
  "error": "Bad Request",
  "message": "Missing username or password",
  "statusCode": 400,
}
`;

exports[`routes/sessions should receive a 401 on GET /me without auth 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`routes/sessions should receive a 403 on POST /login with invalid credentials 1`] = `
Object {
  "error": "Forbidden",
  "message": "Invalid username or password",
  "statusCode": 403,
}
`;
